Split files_sharing middelware

rullzer opened the pull request:
Fixes #19477

Since for external shares there is no need for link shares to be enabled
we should check which controller is actually being called.

This makes sure that in all cases we verify that the files_sharing app
is enabled. But only for the share controller (public shares) we check
if the API is enabled and if links are enabled.

TODO: 
- [x] add checks for federated sharing as well

rullzer committed:
Split files_sharing middelware

Since for external shares there is no need for link shares to be enabled
we should check which controller is actually being called.

This makes sure that in all cases we verify that the files_sharing app
is enabled. But only for the share controller (public shares) we check
if the API is enabled and if links are enabled.

TODO: add checks for federated sharing as well

LukasReschke commented on apps/files_sharing/lib/middleware/sharingcheckmiddleware.php:
```diff
@@ -68,6 +68,13 @@ public function beforeController($controller, $methodName) {
+			//TODO: Proper checks
+		} else if ($controller instanceof \OCA\Files_Sharing\Controllers\ShareController &&
+			$this->isLinkSharingEnabled()) {
+			throw new \Exception('Link sharing is disabled');
```
Can we do a NotFoundException?

LukasReschke replied:
Reason is https://github.com/owncloud/core/blob/d74568dab8b230601f6e63c5fb0c50ae444f218c/apps/files_sharing/lib/middleware/sharingcheckmiddleware.php#L89-L95

rullzer replied:
fixed

rullzer committed:
sharingcheckmiddleware now handles externalshares as well

Added new annotations for the externalsharescontroller class
* @NoOutgoingFederatedSharingRequired
* @NoIncomingFederatedSharingRequired

By default both are required for all functions in the
externalSharesController.

A proper exception is thrown and then a 405 is returned instead of the
default error page. Since it is only an API endpoint this makes more
sense.

Unit tests added and updated

rullzer commented:
Ok the middleware now properly handles the externalshares stuff as well.

* For the externalshares the class uses 2 annotations. By default all methods require incomming and outgoing s2s shares to be enabled. Annotation only disable the check.
* Since it is an API endpoint do not show the generic error page, rather return a 405
* Fix and updated unit tests (coverage of the middleware at 100%)

Review time!

CC: @LukasReschke @MorrisJobke @PVince81 @nickvergessen @DeepDiver1975

schiesbn commented:
great work, code looks good :+1:

PVince81 commented:
Makes sense :+1:

MorrisJobke commented:
@rullzer Is it possible to backport this to stable8.1? https://github.com/owncloud/core/issues/19477#issuecomment-144657239

rullzer commented:
@MorrisJobke how badly do we want to backport this? Since there were multiple changes for 8.2 in those files. I can backport them all in 1 (they are all pretty clean). But only this one is kind of messy.

MorrisJobke commented:
> @MorrisJobke how badly do we want to backport this? Since there were multiple changes for 8.2 in those files. I can backport them all in 1 (they are all pretty clean). But only this one is kind of messy.

@michaelstingl Can you ellaborate on this?

michaelstingl commented:
@MorrisJobke @rullzer A backport to 8.1.4 would be awesome! (Important for the user due to legal
reasons)

rullzer commented:
backport in #19638
